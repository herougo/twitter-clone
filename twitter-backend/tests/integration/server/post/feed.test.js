const createApp = require("../../../../src/server/createApp");
const DI_NAMES = require("../../../../src/server/dependency-injection/names");
const request = require('supertest');
const createMockLogger = require("../../../utils/mocks/mockLogger");
const mongoose = require("mongoose");
const { populateDatabase, clearDatabase } = require("../../../utils/database/changeContents");
const { DB_IDS } = require("../../../utils/database/ids");
const { sanitizePaths } = require("../../../utils/sanitization");
const { USER_JWT_TOKENS } = require("../../../utils/database/userData");

let app;
let diContainer;
let endpoint = '/post/feed';

// run once before all suites in the file
beforeAll(async () => {
    const appData = await createApp({
        [DI_NAMES.logger]: createMockLogger()
    });
    app = appData.app;
    diContainer = appData.diContainer;
});

// run once after all suites in the file
afterAll(async () => {
    await mongoose.connection.close(); // neccessary to avoid a jest error
});

describe("GET /post/feed endpoint", () => {
    // run before each "test"
    beforeEach(async () => {
        await clearDatabase(diContainer);
        await populateDatabase(diContainer);
    });

    const sendToEndpoint = async (token) => {
        if (!token) {
            return await request(app)
                .get(endpoint)
                .send();
        }

        return await request(app)
            .get(endpoint)
            .set('Authorization', `Bearer ${token}`)
            .send();
    };

    test("Success", async () => {
        const response = await sendToEndpoint(USER_JWT_TOKENS.main);
        expect(response.statusCode).toBe(200);
        expect(response.body).toEqual([]);
    });

    test("Success (follower)", async () => {
        const response = await sendToEndpoint(USER_JWT_TOKENS.follower);
        expect(response.statusCode).toBe(200);
        const sanitizedBody = response.body.map(
            post => sanitizePaths(post, ['createdDate', 'replyTo.createdDate'], '')
        );
        expect(sanitizedBody).toMatchSnapshot();
    });

    test("Invalid User", async () => {
        const response = await sendToEndpoint(USER_JWT_TOKENS.missing);
        expect(response.statusCode).toBe(400);
        expect(response.body.errors.message).toEqual("GetPosts: Invalid user");
    });

    test("No JWT token", async () => {
        const response = await sendToEndpoint();
        expect(response.statusCode).toBe(401);
    });
});