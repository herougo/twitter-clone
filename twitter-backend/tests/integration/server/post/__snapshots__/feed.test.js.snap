// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /post/feed endpoint Success (follower) 1`] = `
[
  {
    "author": {
      "id": "675351d698452e4fadc376be",
      "name": "hi there",
      "profilePicPath": null,
      "username": "username",
    },
    "content": "I like dogs.",
    "createdDate": "",
    "id": "675351d698452e4fadc376c3",
    "numDislikes": 0,
    "numLikes": 0,
    "userInteraction": "",
  },
]
`;
